docs/refman/ppr-index.8.pod
#
# mouse:~ppr/src/docs/refman/ppr-index.8.pod
# Copyright 1995--2003, Trinity College Computing Center.
# Written by David Chappell.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#
# * Redistributions of source code must retain the above copyright notice,
# this list of conditions and the following disclaimer.
# 
# * Redistributions in binary form must reproduce the above copyright
# notice, this list of conditions and the following disclaimer in the
# documentation and/or other materials provided with the distribution.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE 
# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF 
# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
# POSSIBILITY OF SUCH DAMAGE.
#
# Last modified 7 March 2003.
#

=head1 NAME

B<ppr-index> - Scan system and create indexes of things PPR can use

=head1 SYNOPSIS

B<ppr-index> [B<--delete>] [B<fonts>] [B<ppds>] [B<filters>]

=head1 DESCRIPTION

This program scans the directories listed in the [fonts] section of
F</etc/ppr/ppr.conf> and builds an index of the fonts it finds in those
directories.  Once fonts are in the index, PPR can download them when
printing if the job contains DSC comments requesting font downloading.

This program scans the directories listed in the [ppds] section of
F</etc/ppr/ppr.conf> and builds an index of the PPD files it finds in those
directories.  The web interface refers to this index when displaying the
list of available PPD files.

This program searches the PATH for programs which can be used by PPR filter
scripts to convert files in various formats to PostScript.  After gathering
information, this program will create usable filter scripts from a
collection of prototype scripts.  If more than one prototype scripts could
be used, it automatically chooses what it considers to be the best one.  If
the necessary programs for a particular prototype filter script can't be
found, no usable script will be created and if there already is one it will
be removed.

You should run this program anytime you install or remove software that
might be used to convert files to PostScript.  Examples include NetPBM,
Acroread, and HTMLDOC.  This program is run automatically once a day from
PPR's crontab.

=head1 FILES

The font database is stored in the file F</var/spool/ppr/fontindex.db>.

The PPD file database is stored in the file F</var/spool/ppr/ppdindex.db>.

The prototype filters scripts are found in F</usr/lib/ppr/fixup/>.  The
usable filter scripts created by this program are deposited in
F</usr/lib/ppr/filters/>.  When this program is run from cron, the output is
logged in F</var/spool/ppr/logs/ppr-indexfilters>.

=head1 SEE ALSO

"PPR, a PostScript Print Spooler"

=head1 HISTORY

PPR was written at Trinity College during 1993--2003.

=head1 AUTHOR

David Chappell, Trinity College Computing Center, Hartford, Connecticut.

=cut


History.txt
mouse:~ppr/src/History.txt (symbolic link)

This file contains a history of changes to PPR which occured after the
release of version 1.00.


1.00, 26 April 1995

First public release.


1.10, 27 April 1995--7 June 1995

Removed "From:" line from message passed to Sendmail by pprd when sending
a message listing the remaining printer faults.  This is so that Sendmail
will take the "real name" from the password file and put it in the "From:"
line.

Removed dotmatrix/input.epson, a GIF file which had been inadvertently
included in the distribution.

Moved the contents of include/atalk_streams.h into
interfaces/atalk_streams.c since it was not long enough to justify an
include file, nor was it of interest to any other program.

Added code to ppad to classify members of the group "ppradmin" as PPR
addministrators who can edit printer and group configurations.

Added -v switch to ppad.

Removed uid's 1-99 from the list of priveledged users.

Fixed ppad(8) so that it considers the group "ppradmin" as priveledged
rather than "ppr".

Changed ppr(1) so that it will accept the members of the group "pprprox" as
priveledged as well as the members of "ppr".  This is for security reasons
as one may not want to grant a user who has the privedge of using the -f
switch the priveledge of deleting other people's jobs.

Modified ppop(1) and pprd(8) so that user's can request to see the queue
entry for a single job.  Previously, the only options were all jobs for a
certain destination or all jobs in all queues.

Rewrote and renamed the undocumented "ppop qdebug" command to become the new
documented command "ppop details".  (This was done primarily to support
wwwppop, an experimental managment interface for PPR.  Since it was still
incomplete it was not released with this version of PPR.)

Renamed interfaces/atalk_streams.c to atalk_ali.c and
papsrv/papsrv_streams.c to papsrv_ali.c

Removed code from "make dirs" so that it doesn't create the directory
"/etc/ppr/groups.ppd".

Added sleep=N and eoftimeout=N options to tcpip interface.

Added -C _filename_ switch to ppr invokation in xppr.

Added "-default okay" to xmessage invokation in xppr.

Added the "ppop lpq" sub-command to the "ppop help" information, thus
making it a documented sub-command.

Extended automatic bin selection to multiple media types when "%%PageMedia:"
comments are present.  This has not yet been well tested.

Made "%%EndProcset" an acceptable substitute for "%%EndProcSet" due to a
reputed bug in QuickDraw GX.

Changed rule in ppr_infile.c for detecting PCL code so that documents with
only one escape would still be detected.


1.20, 8 June 1995--10 January 1996

Fixed code in ppr_infile.c so that the ppr -o switch will work even if there
is no "DefFiltOpts:" line in the printer or group configuration file.

Added '(' to list of likely second characters for HP escape codes
in ppr_infile.c

Replaced printing of errno value with strerror(errno) output in some places,
eliminated printing of errno=x from other places where both were printed.

Fixed ppr(1) so that it sets the environment variables "PATH", "IFS", and
"SHELL" so that filters and responders it invokes are less easy to subvert.

Removed setting of PATH from ~ppr/src/misc_filters/tex.perl and changed
shell name from "/usr/bin/tainperl" to "/usr/bin/perl".

Changed code which determins default responder address so that the
environment variables "USER" and "LOGNAME" are used in preference to the
current uid.

Added prompts in Configure to set the directory into which the links to PPR
binaries will be written and the path to the X-Windows binaries such as
xterm.

Modified papsrv so that it will pass the network and node number of the
client to ppr with the -r switch.  (This is for use with an experimental
AppleTalk responder.)

Modified papsrv.c, renaming cleanup() to termination_handler() and causing
it to call fatal() only the first time it is called.  Without this, name
could be left on the network if the CAP version received two terminate
signals in rapid sucession.

Added code to re-line-wrap responder messages for certain responders.
(The library function responder_width() returns the width for a responder.
Currently it is hard coded to return zero for all responders except "samba".
A return value of zero indicates that the message should not be
re-line-wrapped.)

Dropt reading of LPDEST and PRINTER by ppr(1).  (I can not really remember
why I put this feature in.  I don't think the reasons where very good.  It
may have been that I thought that the same printer would be available thru
two spooling systems so one might want to set the default destination for
both with one variable.  That now seems a little far fetched.)

Separated Solaris and Starserver sections of sysdep.h

Added PNG (Portable Network Graphics) format with magic number recognition
to ppr_infile.c.  This remains untested since I do not have any programs
which manipulate this graphics format.

In pprd.c, put sleep(1) between execl("pprdrv") and _exit(242) so that pprd
has time to record the PID before the child exits.  (If this is not done
and execl() fails immediately, a child which should have become pprdrv may
exit before the parent records its PID in the information structure for that
queue.  This result is that it will not recoginize the purpose of the child
when it receives notification of its death and will leave the queue in the
printing state indefinitely.)

Did same in pprdrv.c in code which executes interface.  Also changed
_exit(255) to _exit(242).  (Exit 242 is the new standard since PERL uses 255
as the exit code when die() is executed.)

Added automatic creation of "DefFiltOpts:" lines to ppad(8).  Manual
creation is supported by the the "ppad deffiltopts" and "ppad group
deffiltopts" command.

Modified ppop so that interactive mode restores the terminal settings when
it exits.  Also changed terminal setting code to use the POSIX method of

cron/ppr-index.c
/*
** mouse:~ppr/src/cron/ppr-index.c
** Copyright 1995--2004, Trinity College Computing Center.
** Written by David Chappell.
**
** Redistribution and use in source and binary forms, with or without
** modification, are permitted provided that the following conditions are met:
**
** * Redistributions of source code must retain the above copyright notice,
** this list of conditions and the following disclaimer.
** 
** * Redistributions in binary form must reproduce the above copyright
** notice, this list of conditions and the following disclaimer in the
** documentation and/or other materials provided with the distribution.
**
** THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
** AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
** IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
** ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE 
** LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
** CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF 
** SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
** INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
** CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
** ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
** POSSIBILITY OF SUCH DAMAGE.
**
** Last modified 19 December 2004.
*/

#include "config.h"
#include <stdio.h>
#include <string.h>
#ifdef INTERNATIONAL
#include <locale.h>
#include <libintl.h>
#endif
#include "gu.h"
#include "global_defines.h"
#include "util_exits.h"
#include "version.h"

const char myname[] = "ppr-index";

const char *indexes[] = {
	"fonts",
	"ppds",
	"filters",
	"windrivers",
	NULL
	};

static int do_index(const char name[], gu_boolean delete_index);

/*
** Command line options:
*/
static const char *option_chars = "";
static const struct gu_getopt_opt option_words[] =
		{
		{"delete", 1000, FALSE},
		{"help", 9000, FALSE},
		{"version", 9001, FALSE},
		{(char*)NULL, 0, FALSE}
		} ;

/*
** Print help.
*/
static void help_usage(FILE *outfile)
	{
	fprintf(outfile, _("Usage: %s [switches]\n"), myname);

	fputc('\n', outfile);

	fputs(_("Valid switches:\n"), outfile);

	fputs(_(	"\t--delete\n"), outfile);

	fputs(_(	"\t--version\n"
				"\t--help\n"), outfile);
	}

int main(int argc, char *argv[])
	{
	gu_boolean opt_delete = FALSE;
	int i;
	int ret = EXIT_OK;

	/* Initialize international messages library. */
	#ifdef INTERNATIONAL
	setlocale(LC_ALL, "");
	bindtextdomain(PACKAGE, LOCALEDIR);
	textdomain(PACKAGE);
	#endif

	/* Parse the options. */
	{
	struct gu_getopt_state getopt_state;
	int optchar;
	gu_getopt_init(&getopt_state, argc, argv, option_chars, option_words);
	while((optchar = ppr_getopt(&getopt_state)) != -1)
		{
		switch(optchar)
			{
			case 1000:					/* --delete */
				opt_delete = TRUE;
				break;

			case 9000:					/* --help */
				help_usage(stdout);
				return EXIT_OK;

			case 9001:					/* --version */
				puts(VERSION);
				puts(COPYRIGHT);
				puts(AUTHOR);
				return EXIT_OK;

			default:					/* other getopt errors or missing case */
				gu_getopt_default(myname, optchar, &getopt_state, stderr);
				return EXIT_SYNTAX;
			}
		}
	i = getopt_state.optind;
	}

	/* Main code continues here */
	if(i < argc)
		{
		for( ;i < argc; i++)
			{
			int i2;
			for(i2 = 0; indexes[i2]; i2++)
				{
				if(strcmp(indexes[i2], argv[i]) == 0)
					break;
				}
			if(!indexes[i2])
				{
				fprintf(stderr, "%s: there is no index called \"%s\"\n", argv[0], argv[i]);
				ret = EXIT_BADDEST;
				break;
				}
			if((ret = do_index(argv[i], opt_delete)) != EXIT_OK)
				{
				break;
				}
			}
		}
	else
		{
		for(i = 0; indexes[i]; i++)
			if((ret = do_index(indexes[i], opt_delete)) != EXIT_OK)
				break;
		}

	if(ret == -1)
		return EXIT_INTERNAL;
	else
		return ret;	   
	} /* end of main() */

static int do_index(const char name[], gu_boolean delete_index)
	{
	char fname[MAX_PPR_PATH];
	ppr_fnamef(fname, "%s/lib/index%s", HOMEDIR, name);
	return gu_runl(myname, stderr, fname, delete_index ? "--delete" : NULL, NULL);
	}

/* end of file */
